### PR TITLE
Fix panic when trying to call decode_numeric with an empty buf

### LINKED ISSUE #3: Panic 'byte index 1 is out of bounds of ``'
Failing test:

```
    #[test]
    fn test_decode_panic() {
        let bytes = b"&#1
 ;";
        escaper::decode_html_buf_sloppy(bytes).ok();
    }
```

```
running 1 test
thread 'mimeparser::tests::test_decode_panic' panicked at 'byte index 1 is out of bounds of ``', /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/escaper-0.1.0/src/decode.rs:168:53
stack backtrace:
   0: rust_begin_unwind
             at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/core/src/panicking.rs:92:14
   2: core::str::slice_error_fail
   3: core::str::traits::<impl core::slice::index::SliceIndex<str> for core::ops::range::RangeFrom<usize>>::index
   4: core::str::traits::<impl core::ops::index::Index<I> for str>::index
   5: <alloc::string::String as core::ops::index::Index<core::ops::range::RangeFrom<usize>>>::index
   6: escaper::decode::decode_html_rw
   7: escaper::decode::decode_html_buf_sloppy
```

### PR DESCRIPTION
**Pull request by Jikstra**
Probably fixes https://github.com/dignifiedquire/rust-escaper/issues/3

I don't really understand the things around this, so let's be careful.

### CODE REVIEW

File: src/decode.rs [hunk 1]
```diff
@@ -267,6 +269,10 @@ fn decode_named_entity(entity: &str) -> Result<&'static str, DecodeErrKind> {
 }
 
 fn decode_numeric(esc: &str, radix: u32) -> Result<char, DecodeErrKind> {
+    if esc.len() < 1 {
+       return Err(InvalidCharacter);
```
Comment (Jikstra) [inline src/decode.rs:271]: I guess there's a better error for this case

File: tests/test.rs [hunk 2]
```diff
@@ -103,3 +103,11 @@ fn random_str<R: rand::Rng>(rng: &mut R) -> String {
     }
     return s;
 }
+
+
+#[test]
+fn test_decode_panic() {
```
Comment (Jikstra) [inline tests/test.rs:109]: And there's probably also a better name for this ;)

File: src/decode.rs [hunk 3]
```diff
@@ -114,6 +114,7 @@ pub fn decode_html_rw<R: BufRead, W: Write>(
             }
             Ok(c) => c,
         };
+        println!("'{}', '{}'", c, buf);
```
Comment (Hocuri) [inline src/decode.rs]: Leftover debug statements?

File: src/decode.rs [hunk 4]
```diff
@@ -165,13 +166,14 @@ pub fn decode_html_rw<R: BufRead, W: Write>(
             Numeric if c == 'x' => state = Hex,
             Dec if c == ';' => {
                 state = Normal;
-                let ch = try_parse!(decode_numeric(&buf[1..], 10), good_pos);
+                println!("xxx {:?} {}", buf, good_pos);
```
Comment (Hocuri) [inline src/decode.rs]: also here

**Review by Hocuri — APPROVED** [review]
TBH I don't actually understand what is being done here or why the first character has to be removed in the first place. This means that I can't propose better names, nor can I say whether an error should be returned at all. I mean, maybe we just need to remove the line break, and then it can be read as totally valid HTML?

But the code itself lgtm :)

**Comment by Hocuri** [comment]
Also, I'm wondering if we should try to get this fix into upstream (`rust-htmlescape`), which leads to the question why @dignifiedquire made a fork.

File: src/decode.rs (hunk 3 above)
Comment (Jikstra) [reply src/decode.rs]: Oh yup :)

**Comment by Jikstra** [comment]
> TBH I don't actually understand what is being done here or why the first character has to be removed in the first place. This means that I can't propose better names, nor can I say whether an error should be returned at all. I mean, maybe we just need to remove the line break, and then it can be read as totally valid HTML?

My current theory is that it's about how to decode html entities. There are different formats on encoding, one of them looks like this `&#8209;`, the number 8209 is decimal, and the same entity in hexadecimal ` &#x2011;`. I guess the slicing is used for skipping the x character, what makes we wonder because for the decimal representation we don't need this.

EDIT: Oh. Maybe that's the bug. We're trying to skip the x on the decimal representation. In your test you're using `&#1;` which is clearly a decimal representation.
EDIT2: We're not trying to skip the x. if `c == 'x'` we just change the State from Numeric to Dec and don't push the character to the buf. But what we skip is the `&`. And we always have to skip it.

**Comment by link2xt** [comment]
@Jikstra I made an alternative fix #5 which I think is more correct/fixes the root issue. I explained at https://github.com/dignifiedquire/rust-escaper/issues/3#issuecomment-835473263

**Pull request closed without merging**